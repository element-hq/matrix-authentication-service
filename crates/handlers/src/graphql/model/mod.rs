// Copyright 2024, 2025 New Vector Ltd.
// Copyright 2022-2024 The Matrix.org Foundation C.I.C.
//
// SPDX-License-Identifier: AGPL-3.0-only
// Please see LICENSE in the repository root for full details.

use async_graphql::{Enum, Interface, Object, SimpleObject};
use chrono::{DateTime, Utc};

mod browser_sessions;
mod compat_sessions;
mod cursor;
mod matrix;
mod node;
mod oauth;
mod site_config;
mod upstream_oauth;
mod users;
mod viewer;

pub use self::{
    browser_sessions::{Authentication, BrowserSession},
    compat_sessions::{CompatSession, CompatSsoLogin},
    cursor::{Cursor, NodeCursor},
    node::{Node, NodeType},
    oauth::{OAuth2Client, OAuth2Session},
    site_config::{SiteConfig, SITE_CONFIG_ID},
    upstream_oauth::{UpstreamOAuth2Link, UpstreamOAuth2Provider},
    users::{AppSession, User, UserEmail, UserRecoveryTicket},
    viewer::{Anonymous, Viewer, ViewerSession},
};

/// An object with a creation date.
#[derive(Interface)]
#[graphql(field(
    name = "created_at",
    desc = "When the object was created.",
    ty = "DateTime<Utc>"
))]
pub enum CreationEvent {
    Authentication(Box<Authentication>),
    CompatSession(Box<CompatSession>),
    BrowserSession(Box<BrowserSession>),
    UserEmail(Box<UserEmail>),
    UserRecoveryTicket(Box<UserRecoveryTicket>),
    UpstreamOAuth2Provider(Box<UpstreamOAuth2Provider>),
    UpstreamOAuth2Link(Box<UpstreamOAuth2Link>),
    OAuth2Session(Box<OAuth2Session>),
}

pub struct PreloadedTotalCount(pub Option<usize>);

#[Object]
impl PreloadedTotalCount {
    /// Identifies the total count of items in the connection.
    async fn total_count(&self) -> Result<usize, async_graphql::Error> {
        self.0
            .ok_or_else(|| async_graphql::Error::new("total count not preloaded"))
    }
}

/// The state of a session
#[derive(Enum, Copy, Clone, Eq, PartialEq)]
pub enum SessionState {
    /// The session is active.
    Active,

    /// The session is no longer active.
    Finished,
}

/// The type of a user agent
#[derive(Enum, Copy, Clone, Eq, PartialEq)]
pub enum DeviceType {
    /// A personal computer, laptop or desktop
    Pc,

    /// A mobile phone. Can also sometimes be a tablet.
    Mobile,

    /// A tablet
    Tablet,

    /// Unknown device type
    Unknown,
}

impl From<mas_data_model::DeviceType> for DeviceType {
    fn from(device_type: mas_data_model::DeviceType) -> Self {
        match device_type {
            mas_data_model::DeviceType::Pc => Self::Pc,
            mas_data_model::DeviceType::Mobile => Self::Mobile,
            mas_data_model::DeviceType::Tablet => Self::Tablet,
            mas_data_model::DeviceType::Unknown => Self::Unknown,
        }
    }
}

/// A parsed user agent string
#[derive(SimpleObject)]
pub struct UserAgent {
    /// The user agent string
    pub raw: String,

    /// The name of the browser
    pub name: Option<String>,

    /// The version of the browser
    pub version: Option<String>,

    /// The operating system name
    pub os: Option<String>,

    /// The operating system version
    pub os_version: Option<String>,

    /// The device model
    pub model: Option<String>,

    /// The device type
    pub device_type: DeviceType,
}

impl From<mas_data_model::UserAgent> for UserAgent {
    fn from(ua: mas_data_model::UserAgent) -> Self {
        Self {
            raw: ua.raw,
            name: ua.name,
            version: ua.version,
            os: ua.os,
            os_version: ua.os_version,
            model: ua.model,
            device_type: ua.device_type.into(),
        }
    }
}
