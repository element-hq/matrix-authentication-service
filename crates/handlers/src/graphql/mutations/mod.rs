// Copyright 2024, 2025 New Vector Ltd.
// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
// Please see LICENSE files in the repository root for full details.

mod browser_session;
mod compat_session;
mod matrix;
mod oauth2_session;
mod user;
mod user_email;
mod user_passkey;

use anyhow::Context as _;
use async_graphql::MergedObject;
use mas_data_model::SiteConfig;
use mas_storage::BoxRepository;
use zeroize::Zeroizing;

use super::Requester;
use crate::passwords::PasswordManager;

/// The mutations root of the GraphQL interface.
#[derive(Default, MergedObject)]
pub struct Mutation(
    user_email::UserEmailMutations,
    user_passkey::UserPasskeyMutations,
    user::UserMutations,
    oauth2_session::OAuth2SessionMutations,
    compat_session::CompatSessionMutations,
    browser_session::BrowserSessionMutations,
    matrix::MatrixMutations,
);

impl Mutation {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }
}

/// Check the password if neeed
///
/// Returns true if password verification is not needed, or if the password is
/// correct. Returns false if the password is incorrect or missing.
async fn verify_password_if_needed(
    requester: &Requester,
    config: &SiteConfig,
    password_manager: &PasswordManager,
    password: Option<String>,
    user: &mas_data_model::User,
    repo: &mut BoxRepository,
) -> Result<bool, async_graphql::Error> {
    // If the requester is admin, they don't need to provide a password
    if requester.is_admin() {
        return Ok(true);
    }

    // If password login is disabled, assume we don't want the user to reauth
    if !config.password_login_enabled {
        return Ok(true);
    }

    // Else we need to check if the user has a password
    let Some(user_password) = repo
        .user_password()
        .active(user)
        .await
        .context("Failed to load user password")?
    else {
        // User has no password, so we don't need to verify the password
        return Ok(true);
    };

    let Some(password) = password else {
        // There is a password on the user, but not provided in the input
        return Ok(false);
    };

    let password = Zeroizing::new(password);

    let res = password_manager
        .verify(
            user_password.version,
            password,
            user_password.hashed_password,
        )
        .await;

    Ok(res.is_ok())
}
