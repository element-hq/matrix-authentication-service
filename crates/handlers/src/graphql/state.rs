// Copyright 2024 New Vector Ltd.
// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// SPDX-License-Identifier: AGPL-3.0-only
// Please see LICENSE in the repository root for full details.

use mas_data_model::SiteConfig;
use mas_matrix::HomeserverConnection;
use mas_policy::Policy;
use mas_router::UrlBuilder;
use mas_storage::{BoxClock, BoxRepository, BoxRng, RepositoryError};

use crate::{graphql::Requester, passwords::PasswordManager};

#[async_trait::async_trait]
pub trait State {
    async fn repository(&self) -> Result<BoxRepository, RepositoryError>;
    async fn policy(&self) -> Result<Policy, mas_policy::InstantiateError>;
    fn password_manager(&self) -> PasswordManager;
    fn homeserver_connection(&self) -> &dyn HomeserverConnection<Error = anyhow::Error>;
    fn clock(&self) -> BoxClock;
    fn rng(&self) -> BoxRng;
    fn site_config(&self) -> &SiteConfig;
    fn url_builder(&self) -> &UrlBuilder;
}

pub type BoxState = Box<dyn State + Send + Sync + 'static>;

pub trait ContextExt {
    fn state(&self) -> &BoxState;

    fn requester(&self) -> &Requester;
}

impl ContextExt for async_graphql::Context<'_> {
    fn state(&self) -> &BoxState {
        self.data_unchecked()
    }

    fn requester(&self) -> &Requester {
        self.data_unchecked()
    }
}
