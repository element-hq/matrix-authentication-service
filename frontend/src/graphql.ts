// Copyright 2024 New Vector Ltd.
// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// SPDX-License-Identifier: AGPL-3.0-only
// Please see LICENSE in the repository root for full details.

import { QueryClient } from "@tanstack/react-query";
import type { ExecutionResult } from "graphql";
import appConfig from "./config";
import type { TypedDocumentString } from "./gql/graphql";

let graphqlEndpoint: string;
if (import.meta.env.TEST && typeof window === "undefined") {
  graphqlEndpoint = new URL(
    appConfig.graphqlEndpoint,
    "http:://localhost/",
  ).toString();
} else {
  graphqlEndpoint = new URL(
    appConfig.graphqlEndpoint,
    window.location.toString(),
  ).toString();
}

type RequestOptions<TData, TVariables> = {
  query: TypedDocumentString<TData, TVariables>;
  signal?: AbortSignal;
  // biome-ignore lint/suspicious/noExplicitAny: this is for inference
} & (TVariables extends Record<any, never>
  ? { variables?: TVariables }
  : { variables: TVariables });

export const graphqlRequest = async <TData, TVariables>({
  query,
  variables,
  signal,
}: RequestOptions<TData, TVariables>): Promise<TData> => {
  let response: Response;
  try {
    response = await fetch(graphqlEndpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        query,
        variables,
      }),
      signal,
    });
  } catch (cause) {
    throw new Error(`GraphQL request to ${graphqlEndpoint} request failed`, {
      cause,
    });
  }

  if (!response.ok) {
    throw new Error(
      `GraphQL request to ${graphqlEndpoint} failed: ${response.status}`,
    );
  }

  const json: ExecutionResult<TData> = await response.json();
  if (json.errors) {
    throw new Error(JSON.stringify(json.errors));
  }

  if (!json.data) {
    throw new Error(`GraphQL request to ${graphqlEndpoint} returned no data`);
  }

  return json.data;
};

export const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      throwOnError: true,
    },
  },
});
