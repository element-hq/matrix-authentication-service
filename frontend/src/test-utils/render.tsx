// Copyright 2024 New Vector Ltd.
//
// SPDX-License-Identifier: AGPL-3.0-only
// Please see LICENSE in the repository root for full details.

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render as testingLibraryRender } from "@testing-library/react";
import { DummyRouter } from "./router";

const client = new QueryClient();

const render = (
  ui: Parameters<typeof testingLibraryRender>[0],
  options: Parameters<typeof testingLibraryRender>[1] = {},
): ReturnType<typeof testingLibraryRender> =>
  testingLibraryRender(ui, {
    wrapper: ({ children }) => (
      <QueryClientProvider client={client}>
        <DummyRouter>{children}</DummyRouter>
      </QueryClientProvider>
    ),
    ...options,
  });

export default render;
