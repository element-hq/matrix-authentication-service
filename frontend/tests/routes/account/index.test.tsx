// Copyright 2024 New Vector Ltd.
//
// SPDX-License-Identifier: AGPL-3.0-only
// Please see LICENSE in the repository root for full details.

// @vitest-environment happy-dom

import { act, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { HttpResponse } from "msw";
import { describe, expect, it } from "vitest";
import { mockSetDisplayNameMutation } from "../../../src/gql/graphql";
import { renderPage, server } from "../render";

describe("Account home page", () => {
  it("renders the page", async () => {
    const { asFragment } = await renderPage("/");
    expect(asFragment()).toMatchSnapshot();
  });

  describe("display name edit box", () => {
    it("lets edit the display name", async () => {
      // TODO: a better way to wait on delays
      let advance: () => void = () => {};
      const wait = new Promise((resolve) => {
        advance = () => resolve(void 0);
      });

      server.use(
        mockSetDisplayNameMutation(async ({ variables: { displayName } }) => {
          await wait;
          // Double check that the display name posted is the one we expect
          if (displayName !== "New display name") {
            throw new Error("Invalid display name");
          }

          return HttpResponse.json({
            data: {
              setDisplayName: {
                __typename: "SetDisplayNamePayload",
                status: "SET",
              },
            },
          });
        }),
      );

      const user = userEvent.setup();
      await renderPage("/");

      // TODO: there is only one 'Edit' button, but we should have better labels
      const editButton = screen.getByLabelText("Edit");
      await user.click(editButton);

      const dialog = screen.getByRole("dialog", { name: "Edit profile" });
      expect(dialog).toBeInTheDocument();
      expect(dialog).toMatchSnapshot();

      const displayNameInput = screen.getByLabelText("Display name");
      const newDisplayName = "New display name";
      await user.clear(displayNameInput);
      await user.type(displayNameInput, newDisplayName);
      const saveButton = screen.getByRole("button", { name: "Save" });
      expect(saveButton).not.toHaveAttribute("aria-disabled", "true");
      await user.click(saveButton);
      expect(saveButton).toHaveAttribute("aria-disabled", "true");
      act(() => advance());
      waitFor(() => expect(dialog).not.toBeInTheDocument());
    });

    it("closes with escape", async () => {
      const user = userEvent.setup();
      await renderPage("/");

      // TODO: there is only one 'Edit' button, but we should have better labels
      const editButton = screen.getByLabelText("Edit");
      await user.click(editButton);

      const dialog = screen.getByRole("dialog", { name: "Edit profile" });
      expect(dialog).toBeInTheDocument();

      await user.keyboard("{esc}");
      waitFor(() => expect(dialog).not.toBeInTheDocument());
    });

    it("closes with cancel", async () => {
      const user = userEvent.setup();
      await renderPage("/");

      // TODO: there is only one 'Edit' button, but we should have better labels
      const editButton = screen.getByLabelText("Edit");
      await user.click(editButton);

      const dialog = screen.getByRole("dialog", { name: "Edit profile" });
      expect(dialog).toBeInTheDocument();

      const cancelButton = screen.getByRole("button", { name: "Cancel" });
      await user.click(cancelButton);
      waitFor(() => expect(dialog).not.toBeInTheDocument());
    });

    it("displays an error if the display name is invalid", async () => {
      server.use(
        mockSetDisplayNameMutation(() =>
          HttpResponse.json({
            data: {
              setDisplayName: {
                __typename: "SetDisplayNamePayload",
                status: "INVALID",
              },
            },
          }),
        ),
      );

      const user = userEvent.setup();
      await renderPage("/");

      // TODO: there is only one 'Edit' button, but we should have better labels
      const editButton = screen.getByLabelText("Edit");
      await user.click(editButton);

      const dialog = screen.getByRole("dialog", { name: "Edit profile" });
      expect(dialog).toBeInTheDocument();

      const displayNameInput = screen.getByRole("textbox", {
        name: "Display name",
      });
      await user.clear(displayNameInput);
      await user.type(displayNameInput, "Something");
      const saveButton = screen.getByRole("button", { name: "Save" });
      await user.click(saveButton);

      expect(displayNameInput).toBeInvalid();
      expect(dialog).toMatchSnapshot();
    });
  });
});
