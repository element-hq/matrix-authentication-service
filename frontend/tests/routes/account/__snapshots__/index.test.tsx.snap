// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Account home page > display name edit box > displays an error if the display name is invalid 1`] = `
<div
  aria-describedby="radix-:r41:"
  aria-labelledby="radix-:r40:"
  class="_body_9cf7b0"
  data-state="open"
  id="radix-:r3v:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <h2
    class="_title_9cf7b0"
    id="radix-:r40:"
  >
    Edit profile
  </h2>
  <span
    aria-label="@alice:example.com"
    class="_avatar_mcap2_17 self-center _avatar-imageless_mcap2_61"
    data-color="6"
    data-type="round"
    role="img"
    style="--cpd-avatar-size: 88px;"
  >
    A
  </span>
  bound HTMLFormElement {
    "0": <input
      aria-describedby="radix-:r4m:"
      aria-invalid="true"
      autocomplete="name"
      class="_control_9gon8_18 _control_1qov4_22"
      data-invalid="true"
      id="radix-:r4f:"
      name="displayname"
      title=""
      type="text"
      value="Alice"
    />,
    "1": <button
      aria-controls=":r4g:"
      aria-labelledby=":r4h:"
      class="_action_1qov4_33"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
        />
      </svg>
    </button>,
    "2": <input
      class="_control_9gon8_18"
      id="radix-:r4n:"
      name="mxid"
      readonly=""
      title=""
      value="@alice:example.com"
    />,
    "3": <button
      aria-disabled="false"
      class="_button_i91xf_17"
      data-kind="primary"
      data-size="lg"
      role="button"
      tabindex="0"
      type="submit"
    >
      Save
    </button>,
  }
  <button
    class="_button_i91xf_17"
    data-kind="tertiary"
    data-size="lg"
    role="button"
    tabindex="0"
    type="button"
  >
    Cancel
  </button>
  <button
    aria-labelledby=":r4o:"
    class="_close_9cf7b0"
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
      />
    </svg>
  </button>
</div>
`;

exports[`Account home page > display name edit box > lets edit the display name 1`] = `
<div
  aria-describedby="radix-:ro:"
  aria-labelledby="radix-:rn:"
  class="_body_9cf7b0"
  data-state="open"
  id="radix-:rm:"
  role="dialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <h2
    class="_title_9cf7b0"
    id="radix-:rn:"
  >
    Edit profile
  </h2>
  <span
    aria-label="@alice:example.com"
    class="_avatar_mcap2_17 self-center _avatar-imageless_mcap2_61"
    data-color="6"
    data-type="round"
    role="img"
    style="--cpd-avatar-size: 88px;"
  >
    A
  </span>
  bound HTMLFormElement {
    "0": <input
      aria-describedby="radix-:r1d:"
      autocomplete="name"
      class="_control_9gon8_18 _control_1qov4_22"
      id="radix-:r16:"
      name="displayname"
      title=""
      type="text"
      value="Alice"
    />,
    "1": <button
      aria-controls=":r17:"
      aria-labelledby=":r18:"
      class="_action_1qov4_33"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
        />
      </svg>
    </button>,
    "2": <input
      class="_control_9gon8_18"
      id="radix-:r1e:"
      name="mxid"
      readonly=""
      title=""
      value="@alice:example.com"
    />,
    "3": <button
      aria-disabled="false"
      class="_button_i91xf_17"
      data-kind="primary"
      data-size="lg"
      role="button"
      tabindex="0"
      type="submit"
    >
      Save
    </button>,
  }
  <button
    class="_button_i91xf_17"
    data-kind="tertiary"
    data-size="lg"
    role="button"
    tabindex="0"
    type="button"
  >
    Cancel
  </button>
  <button
    aria-labelledby=":r1f:"
    class="_close_9cf7b0"
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.293 6.293a1 1 0 0 1 1.414 0L12 10.586l4.293-4.293a1 1 0 1 1 1.414 1.414L13.414 12l4.293 4.293a1 1 0 0 1-1.414 1.414L12 13.414l-4.293 4.293a1 1 0 0 1-1.414-1.414L10.586 12 6.293 7.707a1 1 0 0 1 0-1.414Z"
      />
    </svg>
  </button>
</div>
`;

exports[`Account home page > renders the page 1`] = `
<DocumentFragment>
  <div
    class="_layoutContainer_0c8bf9 _wide_0c8bf9"
  >
    <div
      class="flex flex-col gap-4"
    >
      <header
        class="flex justify-between mb-4"
      >
        <h1
          class="_typography_yh5dq_162 _font-heading-lg-semibold_yh5dq_135"
        >
          Your account
        </h1>
        <button
          aria-controls="radix-:r0:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="_button_i91xf_17 _has-icon_i91xf_66 _destructive_i91xf_116"
          data-kind="secondary"
          data-size="sm"
          data-state="closed"
          role="button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 12.031c0-.283.096-.52.288-.712A.968.968 0 0 1 10 11.03h7.15l-1.875-1.875a.96.96 0 0 1-.3-.7c0-.266.108-.508.325-.725a.93.93 0 0 1 .712-.287.977.977 0 0 1 .688.287l3.6 3.6c.1.1.17.209.212.325.042.117.063.242.063.375 0 .134-.02.259-.063.375a.877.877 0 0 1-.212.325l-3.6 3.6a.93.93 0 0 1-.712.288.977.977 0 0 1-.688-.288 1.02 1.02 0 0 1-.313-.712.931.931 0 0 1 .288-.713l1.875-1.875H10a.968.968 0 0 1-.712-.287A.968.968 0 0 1 9 12.03Zm-6-7c0-.55.196-1.02.588-1.412A1.926 1.926 0 0 1 5 3.03h6a.97.97 0 0 1 .713.288.968.968 0 0 1 .287.712.97.97 0 0 1-.287.713.968.968 0 0 1-.713.287H5v14h6a.97.97 0 0 1 .713.288.968.968 0 0 1 .287.712.97.97 0 0 1-.287.713.968.968 0 0 1-.713.287H5c-.55 0-1.02-.196-1.412-.587A1.926 1.926 0 0 1 3 19.03v-14Z"
            />
          </svg>
          Sign out
        </button>
      </header>
      <div
        class="_user_66f22a"
      >
        <span
          aria-label="@alice:example.com"
          class="_avatar_mcap2_17 _avatar_66f22a _avatar-imageless_mcap2_61"
          data-color="6"
          data-type="round"
          role="img"
          style="--cpd-avatar-size: var(--cpd-space-14x);"
        >
          A
        </span>
        <div
          class="_meta_66f22a"
        >
          <p
            class="_typography_yh5dq_162 _font-body-lg-semibold_yh5dq_83"
          >
            Alice
          </p>
          <p
            class="_typography_yh5dq_162 _font-body-md-regular_yh5dq_59 _mxid_66f22a"
          >
            @alice:example.com
          </p>
        </div>
        <button
          aria-controls="radix-:r3:"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-labelledby=":r6:"
          class="_icon-button_bh2qc_17 _editButton_66f22a"
          data-state="closed"
          role="button"
          style="--cpd-icon-button-size: var(--cpd-space-6x);"
          tabindex="0"
          type="button"
        >
          <div
            class="_indicator-icon_133tf_26"
            style="--cpd-icon-button-size: 100%;"
          >
            <svg
              fill="currentColor"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M15.706 2.637a2 2 0 0 1 2.829 0l2.828 2.828a2 2 0 0 1 0 2.829L9.605 20.052a1 1 0 0 1-.465.263L3.483 21.73a1 1 0 0 1-1.212-1.212l1.414-5.657a1 1 0 0 1 .263-.465L15.706 2.637Zm1.224 7.262L14.102 7.07l-8.544 8.544-.943 3.77 3.771-.942L16.93 9.9Z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </button>
      </div>
      <nav
        class="_navBar_214497"
      >
        <ul
          class="_navBarItems_214497"
        >
          <li
            class="_navTab_8603fc"
          >
            <a
              aria-current="page"
              class="_navItem_8603fc"
              data-status="active"
              href="/"
            >
              Settings
            </a>
          </li>
          <li
            class="_navTab_8603fc"
          >
            <a
              class="_navItem_8603fc"
              href="/sessions"
            >
              Devices
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      class="_blockList_f8cc7f"
    >
      <div
        class="flex flex-col gap-4"
        id="emails"
      >
        bound HTMLFormElement {
          "0": <input
            aria-describedby="radix-:rc:"
            class="_control_9gon8_18 _userEmailField_e2a518"
            id="radix-:rb:"
            name="email"
            readonly=""
            title=""
            type="email"
            value="alice@example.com"
          />,
        }
        <div
          aria-busy="true"
          class="self-center _mini_0c7436"
          role="alert"
        >
          <svg
            class="_loadingSpinnerInner_0c7436"
            fill="none"
            role="img"
            viewBox="0 0 100 101"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Loading…
            </title>
            <path
              d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
              fill="currentFill"
            />
          </svg>
        </div>
        bound HTMLFormElement {
          "0": <input
            aria-describedby="radix-:re:"
            class="_control_9gon8_18"
            id="radix-:rd:"
            name="input"
            required=""
            title=""
            type="email"
          />,
        }
      </div>
      <div
        class="_separator_144s5_17"
        data-kind="primary"
        data-orientation="horizontal"
        role="separator"
      />
      bound HTMLFormElement {
        "0": <input
          aria-describedby="radix-:rg:"
          class="_control_9gon8_18"
          id="radix-:rf:"
          name="password_preview"
          readonly=""
          title=""
          type="password"
          value="this looks like a password"
        />,
      }
      <div
        class="_separator_144s5_17"
        data-kind="primary"
        data-orientation="horizontal"
        role="separator"
      />
      <div
        data-state="closed"
      >
        <button
          aria-controls="radix-:rh:"
          aria-expanded="false"
          class="_trigger_f1daaa"
          data-state="closed"
          type="button"
        >
          <div
            class="_triggerTitle_f1daaa"
          >
            <h1
              class="_typography_yh5dq_162 _font-heading-sm-semibold_yh5dq_102"
            >
              End-to-end encryption
            </h1>
          </div>
          <svg
            class="_triggerIcon_f1daaa"
            fill="currentColor"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m12 10.775-3.9 3.9a.948.948 0 0 1-.7.275.948.948 0 0 1-.7-.275.948.948 0 0 1-.275-.7.95.95 0 0 1 .275-.7l4.6-4.6c.1-.1.208-.17.325-.213.117-.041.242-.062.375-.062s.258.02.375.062a.877.877 0 0 1 .325.213l4.6 4.6a.948.948 0 0 1 .275.7.949.949 0 0 1-.275.7.948.948 0 0 1-.7.275.948.948 0 0 1-.7-.275l-3.9-3.9Z"
            />
          </svg>
        </button>
        <div
          class="_content_f1daaa"
          data-state="closed"
          hidden=""
          id="radix-:rh:"
        />
      </div>
    </div>
    <footer
      class="_legalFooter_eb428f"
    >
      <nav>
        <a
          class="_link_1mzip_17"
          data-kind="primary"
          href="https://matrix.org/policy"
          rel="noreferrer noopener"
          title="Link to the service privacy policy"
        >
          Privacy Policy
        </a>
        <div
          aria-hidden="true"
          class="_separator_eb428f"
        >
          •
        </div>
        <a
          class="_link_1mzip_17"
          data-kind="primary"
          href="https://matrix.org/tos"
          rel="noreferrer noopener"
          title="Link to the service terms and conditions"
        >
          Terms & Conditions
        </a>
      </nav>
      <p
        class="_imprint_eb428f"
      >
        All Rights Reserved. The Super Chat name, logo and device are registered trade marks of BigCorp Ltd.
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
